File: src/test/java/com/freya02/botcommands/test/commands/text/TextTest.java
package com.freya02.botcommands.test.commands.text;

import com.freya02.botcommands.api.commands.annotations.BotPermissions;
import com.freya02.botcommands.api.commands.annotations.Optional;
import com.freya02.botcommands.api.commands.annotations.UserPermissions;
import com.freya02.botcommands.api.commands.prefixed.BaseCommandEvent;
import com.freya02.botcommands.api.commands.prefixed.CommandEvent;
import com.freya02.botcommands.api.commands.prefixed.TextCommand;
import com.freya02.botcommands.api.commands.prefixed.annotations.JDATextCommand;
import com.freya02.botcommands.api.commands.prefixed.annotations.TextOption;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.User;
import org.jetbrains.annotations.Nullable;

@UserPermissions(Permission.ADMINISTRATOR)
@BotPermissions(Permission.MANAGE_SERVER)
public class TextTest extends TextCommand {
	@JDATextCommand(
			name = "test"
	)
	public void run(BaseCommandEvent event,
	                @TextOption User user,
					@TextOption @Optional long delDays,
	                @TextOption @Nullable String reason
	) {
		event.replyFormat("ok %s : %s : %s", user, delDays, reason).queue();
	}

	@BotPermissions(value = Permission.BAN_MEMBERS, append = true)
	@JDATextCommand(
			name = "test",
			subcommand = "lol",
			order = 1
	)
	public void run(BaseCommandEvent event, @TextOption long l) {
		event.reply("no2").queue();
	}

	@BotPermissions(value = Permission.KICK_MEMBERS, append = true)
	@JDATextCommand(
			name = "test",
			subcommand = "lol",
			order = 2
	)
	public void run(BaseCommandEvent event, @TextOption String xd) {
		event.reply("no").queue();
	}

	@BotPermissions(value = Permission.KICK_MEMBERS, append = true)
	@JDATextCommand(
			name = "test",
			subcommand = "lol"
	)
	public void run(CommandEvent event) {
		event.reply("no3").queue();
	}
}

File: src/test/kotlin/com/freya02/botcommands/test/commands_kt/text/TextTest.kt
package com.freya02.botcommands.test.commands_kt.text

import com.freya02.botcommands.api.BContext
import com.freya02.botcommands.api.annotations.CommandMarker
import com.freya02.botcommands.api.commands.CommandPath
import com.freya02.botcommands.api.commands.annotations.GeneratedOption
import com.freya02.botcommands.api.commands.prefixed.BaseCommandEvent
import com.freya02.botcommands.api.commands.prefixed.TextCommand
import com.freya02.botcommands.api.commands.prefixed.TextCommandManager
import com.freya02.botcommands.api.commands.prefixed.TextGeneratedValueSupplier
import com.freya02.botcommands.api.commands.prefixed.annotations.JDATextCommand
import com.freya02.botcommands.api.commands.prefixed.annotations.TextDeclaration
import com.freya02.botcommands.api.commands.prefixed.annotations.TextOption
import com.freya02.botcommands.api.parameters.ParameterType

@CommandMarker
class TextTest : TextCommand() {
    override fun getGeneratedValueSupplier(
        commandPath: CommandPath,
        optionName: String,
        parameterType: ParameterType
    ): TextGeneratedValueSupplier {
        if (optionName ==  "user_name") {
            return TextGeneratedValueSupplier {
                event -> event.author.name
            }
        }

        return super.getGeneratedValueSupplier(commandPath, optionName, parameterType)
    }

    @JDATextCommand(name = "test_annotated")
    fun onTextTestFallback(
        event: BaseCommandEvent,
        context: BContext,
        @GeneratedOption userName: String
    ) {
        event.reply("""
            text: [fallback]
            context: $context
            user name: $userName
        """.trimIndent()).queue()
    }

    @JDATextCommand(name = "test_annotated")
    fun onTextTest(
        event: BaseCommandEvent,
        @TextOption text: String,
        context: BContext,
        @GeneratedOption userName: String
    ) {
        event.reply("""
            text: $text
            context: $context
            user name: $userName
        """.trimIndent()).queue()
    }

    @JDATextCommand(name = "test_annotated", subcommand = "subcommand")
    fun onTextTestSubcommand(
        event: BaseCommandEvent,
        @TextOption number: Double
    ) {
        event.reply("""
            number: $number
        """.trimIndent()).queue()
    }

    @TextDeclaration
    fun declare(textCommandManager: TextCommandManager) {
        textCommandManager.textCommand("test") {
            option("text")

            customOption("context")

            generatedOption("userName") {
                it.author.name
            }

            function = ::onTextTest
        }

        textCommandManager.textCommand("test") {
            customOption("context")

            generatedOption("userName") {
                it.author.name
            }

            function = ::onTextTestFallback
        }

        textCommandManager.textCommand("test", subcommand = "subcommand") {
            option("number")

            function = ::onTextTestSubcommand
        }
    }
}